R-package/configure

```diff
@@ -1,6 +1,6 @@
 #! /bin/sh
 # Guess values for system-dependent variables and create Makefiles.
-# Generated by GNU Autoconf 2.69 for lightgbm 3.1.0.99.
+# Generated by GNU Autoconf 2.69 for lightgbm 3.1.1.
```

Comment:
/gha run r-valgrind

Reply:
this will probably take 40-60 minutes: https://github.com/microsoft/LightGBM/actions/runs/402978050

Reply:
I'll use this comment thread for all R-related issues so this PR doesn't pick up too many comments and we can resolve it all at once.

good news: tests are passing on Solaris and on Windows 32-bit!
bad news: there is a `NOTE` from `R CMD CHECK` that I think was introduced by the work for #3390 

> * checking R code for possible problems ... NOTE
  partial argument match of 'data' to 'dataset'
lgb.cv : <anonymous>: warning in getinfo(data = data, name =
  "init_score"): partial argument match of 'data' to 'dataset'
lgb.cv : <anonymous>: warning in getinfo(data = data, name = "weight"):

I'll open a separate PR to fix this and to make sure we catch it in CI. Should be quick.

Reply:
I can see this NOTE in recent builds in CI: https://github.com/microsoft/LightGBM/runs/1502651682

it wasn't caught because the "CRAN incoming note" is no longer thrown now that the package is on CRAN.

https://github.com/microsoft/LightGBM/runs/1502651682

Reply:
`valgrind` looks happy!

```text
==2062== Memcheck, a memory error detector
==2062== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==2062== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==2062== Command: /usr/local/RDvalgrind/lib/R/bin/exec/R --no-readline --vanilla -f testthat.R
==2062== 
Loading required package: R6
==2062== Conditional jump or move depends on uninitialised value(s)
==2062==    at 0x49CF138: gregexpr_Regexc (grep.c:2439)
==2062==    by 0x49D1F13: do_regexpr (grep.c:3100)
==2062==    by 0x49A0058: bcEval (eval.c:7121)
==2062==    by 0x498B67F: Rf_eval (eval.c:727)
==2062==    by 0x498E414: R_execClosure (eval.c:1895)
==2062==    by 0x498E0C7: Rf_applyClosure (eval.c:1821)
==2062==    by 0x499FC8C: bcEval (eval.c:7089)
==2062==    by 0x498B67F: Rf_eval (eval.c:727)
==2062==    by 0x498B1CB: forcePromise (eval.c:555)
==2062==    by 0x49963AB: FORCE_PROMISE (eval.c:5142)
==2062==    by 0x4996566: getvar (eval.c:5183)
==2062==    by 0x499D1A5: bcEval (eval.c:6873)
==2062==  Uninitialised value was created by a stack allocation
==2062==    at 0x49CEC37: gregexpr_Regexc (grep.c:2369)
==2062== 
==2062== 
==2062== HEAP SUMMARY:
==2062==     in use at exit: 321,320,729 bytes in 55,930 blocks
==2062==   total heap usage: 2,659,604 allocs, 2,603,674 frees, 6,003,459,320 bytes allocated
==2062== 
==2062== 336 bytes in 1 blocks are possibly lost in loss record 152 of 2,701
==2062==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2062==    by 0x40149CA: allocate_dtv (dl-tls.c:286)
==2062==    by 0x40149CA: _dl_allocate_tls (dl-tls.c:532)
==2062==    by 0x5702322: allocate_stack (allocatestack.c:622)
==2062==    by 0x5702322: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
==2062==    by 0x56D0DDA: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==2062==    by 0x56C88E0: GOMP_parallel (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==2062==    by 0x1543D7C4: LGBM_DatasetCreateFromCSC (c_api.cpp:1286)
==2062==    by 0x1546BEAB: LGBM_DatasetCreateFromCSC_R (lightgbm_R.cpp:91)
==2062==    by 0x4941E2F: R_doDotCall (dotcode.c:634)
==2062==    by 0x494CCC6: do_dotcall (dotcode.c:1281)
==2062==    by 0x499FB01: bcEval (eval.c:7078)
==2062==    by 0x498B67F: Rf_eval (eval.c:727)
==2062==    by 0x498E414: R_execClosure (eval.c:1895)
==2062== 
==2062== LEAK SUMMARY:
==2062==    definitely lost: 0 bytes in 0 blocks
==2062==    indirectly lost: 0 bytes in 0 blocks
==2062==      possibly lost: 336 bytes in 1 blocks
==2062==    still reachable: 321,320,393 bytes in 55,929 blocks
==2062==                       of which reachable via heuristic:
==2062==                         newarray           : 4,264 bytes in 1 blocks
==2062==         suppressed: 0 bytes in 0 blocks
==2062== Reachable blocks (those to which a pointer was found) are not shown.
==2062== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==2062== 
==2062== For lists of detected and suppressed errors, rerun with: -s
==2062== ERROR SUMMARY: 7 errors from 2 contexts (suppressed: 0 from 0)
```

Reply:
moved discussion on the 'partial argument mismatch' to https://github.com/microsoft/LightGBM/issues/3629 and added a release checklist item for it. I think this issue would prevent the R package from acceptance by CRAN.

Reply:
/gha run r-valgrind

Reply:
https://github.com/microsoft/LightGBM/runs/1508978209?check_suite_focus=true

Reply:
/gha run r-valgrind

Reply:
https://github.com/microsoft/LightGBM/actions/runs/405788688

Reply:
Valgrind is happy ✔️

Reply:
just kicked off R Hub builds for Solaris.

```sh
docker run -v $(pwd):/opt/LightGBM -w /opt/LightGBM -it rocker/verse:latest /bin/bash

R
```

```r
install.packages("rhub")

package_tarball <- paste0("lightgbm_", readLines("VERSION.txt")[1], ".tar.gz")
rhub::check(
    path = package_tarball
    , email = "jaylamb20@gmail.com"
    , check_args = "--as-cran"
    , platform = c(
        "solaris-x86-patched"
        , "solaris-x86-patched-ods"
    )
    , env_vars = c(
        "R_COMPILE_AND_INSTALL_PACKAGES" = "always"
    )
)
```

Reply:
Solaris looks good!

Reply:
@guolinke @StrikerRUS I think this release is ready! All the PRs we wanted to get in have been merged, and the extra CRAN checks for the R package are passing.

Should I submit to CRAN?

Reply:
/gha build r-artifacts

Reply:
https://github.com/microsoft/LightGBM/runs/1512517160?check_suite_focus=true

Reply:
produced this artifact: https://github.com/microsoft/LightGBM/suites/1623533888/artifacts/29854422

I'll wait to submit until @guolinke says it's ok

Reply:
Just submitted to CRAN!